src/coreclr/jit/hwintrinsiclistarm64sve.h

```diff
@@ -237,6 +246,7 @@ HARDWARE_INTRINSIC(Sve,           Scatter32BitNarrowing,
 HARDWARE_INTRINSIC(Sve,           Scatter32BitWithByteOffsetsNarrowing,                             -1,      -1,     false, {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_st1w,       INS_sve_st1w,       INS_invalid,        INS_invalid},     HW_Category_MemoryStore,           HW_Flag_Scalable|HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialCodeGen|HW_Flag_ExplicitMaskedOperation|HW_Flag_LowMaskedOperation)
 HARDWARE_INTRINSIC(Sve,           Scatter8BitNarrowing,                                             -1,      -1,     false, {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_st1b,       INS_sve_st1b,       INS_sve_st1b,       INS_sve_st1b,       INS_invalid,        INS_invalid},     HW_Category_MemoryStore,           HW_Flag_Scalable|HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialCodeGen|HW_Flag_ExplicitMaskedOperation|HW_Flag_LowMaskedOperation)
 HARDWARE_INTRINSIC(Sve,           Scatter8BitWithByteOffsetsNarrowing,                              -1,      -1,     false, {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_st1b,       INS_sve_st1b,       INS_sve_st1b,       INS_sve_st1b,       INS_invalid,        INS_invalid},     HW_Category_MemoryStore,           HW_Flag_Scalable|HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialCodeGen|HW_Flag_ExplicitMaskedOperation|HW_Flag_LowMaskedOperation)
+HARDWARE_INTRINSIC(Sve,           SetFfr,                                                           -1,      1,      true,  {INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_invalid,        INS_invalid},     HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_ExplicitMaskedOperation|HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialSideEffectMask|HW_Flag_SpecialCodeGen)
```

Comment:
Are you sure `wrffr` is the right instruction to use? I think we should rather be exposing the `setffr` instruction: https://developer.arm.com/documentation/ddi0596/2020-12/SVE-Instructions/SETFFR--Initialise-the-first-fault-register-to-all-true-?lang=en

I do not think the JIT should need to be emitting `wrffr` at all. From what I can tell, `wrffr` would be necessary mainly in the kernel around context switching. The manual says: https://developer.arm.com/documentation/ddi0596/2020-12/SVE-Instructions/WRFFR--Write-the-first-fault-register-?lang=en

> Read the source predicate register and place in the first-fault register (FFR). This instruction is intended to restore a saved FFR and is not recommended for general use by applications.
> 
> This instruction requires that the source predicate contains a MONOTONIC predicate value, in which starting from bit 0 there are zero or more 1 bits, followed only by 0 bits in any remaining bit positions. If the source is not a monotonic predicate value, then the resulting value in the FFR will be UNPREDICTABLE. It is not possible to generate a non-monotonic value in FFR when using SETFFR followed by first-fault or non-fault loads.

 For the JIT I think we should do this:
1. Allow the user to emit `setffr` to get `ffr` into a known state before the first-faulting loads
2. Always emit `readffr` after `LoadVectorFirstFaulting` as you are already doing, moving its value into the created local. DCE should be able to remove it if that local does not get read.
3. Emit `readffr` into the local we have after `setffr`, in case the user is reading the "all true" value for some reason
4. Read directly from the local as part of the `GetFfr` functions, emitting no `rdffr` instructions as part of that intrinsic

I think this will allow getting rid of all the special handling around spills.

The only question I have is what the expected behavior is when the user writes code like:
```csharp
SetFfr();
NestedCallThatMayEmitFirstFaultLoads();
GetFfr();
```

I think with the approach in this PR (and the approach I suggest here), the `ffr` will reflect the "all-true" value that `setffr` sets up. But some users may expect the `ffr` to reflect what happened inside the nested call, since `ffr` is a shared resource in the ISA.

We should perhaps document somewhere that the state of `ffr` is not guaranteed to reflect what happened inside nested calls.

Reply:
BTW, I also think it would be just reasonable to not try to save/restore `FFR` around calls in the way we are doing now. These intrinsics really allow the user directly to interact with a resource that is shared in the CPU, so I do not see why we shouldn't just reflect that. That is, we would have `SetFfr()` intrinsic that emits the `setffr` instruction, plus the `GetFfr` intrinsics to emit the `rdffr` instruction. It would be up to the user to make sure they read it with `GetFfr*` intrinsics after doing their first-faulting loads, before the shared resource may be overwritten by e.g. child calls.

Reply:
> Why not expose setffr instead of wrffr

I asked this https://github.com/dotnet/runtime/pull/104502#discussion_r1677296041 and we need to revisit `SetFfr()` in API review to see if we can just not pass `mask` parameter, unless @tannergooding knows a use case for having a parameter to `SetFfr()`.

> It would be up to the user to make sure they read it with GetFfr* intrinsics after doing their first-faulting loads, before the shared resource may be overwritten by e.g. child calls. 

This was discussed extensively offline before we came to the consensus of doing it that way. But given the fact that we are exposing `GetFfr()` to the user, we want to make sure the state of FFR is preserved after `LoadFF()` because they will use the FFR contents to see which lanes were read successfully.

Reply:
Thanks, I read the internal email thread and that convinced me that the approach here (to model `FFR` more explicitly per method) is the right one.

Reply:
It is very explicit that `wrffr` is available for use in user code and that it mutates the register fully. Particularly in C/C++, this is the only way for a developer to utilize FFR and make a call while preserving the state (they are expected to do the saving/restore across the call boundary themselves).

In .NET, we're making it implicit caller save instead so that the JIT can still do its job without incurring unnecessary expense. But that still leaves the need that developers be able to pass back arbitrary FFR state from a helper method and allow the caller to then set FFR to that returned value.

Reply:
It seems like we should expose `wrffr` and `setffr` separately then? The manual clearly directs users to use `setffr`, so it's odd to me that we don't even expose it (in the no-argument version).

Reply:
We opted to not expose it because the JIT can recognize `WriteFfr(AllBitsSet)` and optimize it to emit `setffr`.

Due to the sheer number of intrinsics that exist we've tried to balance things a little between core functionality and what can be reasonably recognized as a simple IR pattern. We don't expose overloads for all the masked variants of instructions for the same reason, instead recognizing `ConditionalSelect(mask, M(...), mergeValue)`, which has let us cut down on the surface area by several thousand APIs and centralize the support via fairly minimal logic in the JIT instead.

`setffr` is something we still could directly expose, but it's also trivial for the JIT to say `if (op1->IsVectorAllBitsSet()) { MakeSrcContained(node, op1); }` and then `if (op1->IsVectorAllBitsSet()) { ins = INS_setffr; }`